src/Covenant/Expr.hs

```diff
@@ -308,32 +333,41 @@ toExprGraph (ExprBuilder comp) = do
 -- | Construct a literal (constant) value.
 --
 -- @since 1.0.0
-lit :: Natural -> ExprBuilder Id
-lit = idOf . Lit
+lit :: Natural -> ExprBuilder TypedId
+lit n = withType ExprTypeNat <$> idOf (Lit n)
 
 -- | Construct a primitive function call.
 --
 -- @since 1.0.0
-prim :: PrimCall -> ExprBuilder Id
-prim = idOf . Prim
+prim :: PrimCall -> ExprBuilder TypedId
+prim p = withType (error "todo") <$> idOf (Prim p)
 
 -- | Construct a function application. The first argument is (an expression
 -- evaluating to) a function, the second argument is (an expression evaluating
 -- to) an argument.
---
--- = Important note
---
--- Currently, this does not verify that the first argument is indeed a function,
--- nor that the second argument is appropriate.
-app :: Ref -> Ref -> ExprBuilder Id
-app f x = idOf (App f x)
+app :: Ref -> Ref -> ExprBuilder TypedId
+app f x =
+  let argTy = typeOf x
+      funTy = typeOf f
+      resTy = case funTy of
+        ExprTypeLam argTy' bodyTy | argTy' == argTy -> bodyTy
+        ExprTypeLam argTy' bodyTy ->
+          typeError $
+            "Can't apply a function ("
+              ++ (show argTy')
+              ++ " -> "
+              ++ (show bodyTy)
+              ++ ") to an argument "
+              ++ (show argTy)
+        _ -> typeError $ "Can't apply a value which is not a function (" ++ (show funTy) ++ ")"
+   in withType resTy <$> idOf (App f x)
 
 -- | A proof of how many arguments are available to a Covenant program. Put
 -- another way, a value of type @'Scope' n@ means that we are under @n@ lambdas,
 -- each with an argument we can refer to.
 --
 -- @since 1.0.0
-data Scope (n :: Natural) = Scope
+data Scope (n :: Natural) = Scope [ExprType]
```

Comment:
@kozross I spend sometime trying to wrangle a typelevel list, but as you said, it was unpleasant. I think that this is a good compromise. What do you think?

See also `pushScope` below.